### DISCUSSION TITLE
OpenAPI Client Generator issues using Directus OpenAPI Spec

### DISCUSSION THREAD

**Discussion by CoenWarmer** [directus/directus#14426] [Archive]
Hey guys,

We recently upgraded to the latest version of the [OpenAPI Client Generator package](https://github.com/OpenAPITools/openapi-generator-cli) (`2.5.1`, using generator `6.0.1`), and then tried to generate a `typescript-axios` client from Directus' OpenAPI Spec.

After this upgrade there were some issues in the generation process:
<img width="687" alt="Screenshot 2022-07-14 at 11 46 51" src="https://user-images.githubusercontent.com/535564/178954515-a663421d-ed56-43e2-ad31-2e3df5a63aa9.png">

<img width="1291" alt="Screenshot 2022-07-14 at 11 48 21" src="https://user-images.githubusercontent.com/535564/178954792-c4b0d75e-35ae-4db4-a227-77145ef214e8.png">

<img width="1405" alt="Screenshot 2022-07-14 at 11 49 22" src="https://user-images.githubusercontent.com/535564/178954994-c960a1d9-59e5-4cd2-be89-bc772be59d5d.png">

The last screenshot looks quite bad, but actually the client code *is* generated and seems to work, apart from `XMetadata` and `Query` not being found by the generator.

Once I saved the Directus spec to a file and I moved `XMetadata` and `Query` inside the `/components/schemas` section, they both worked.

Now, I could open an issue for this at Directus, however I feel that may not be completely fair as the the responsibility of interpreting the spec also lies at the OpenAPI Client Generator project.

So I wanted to ask you how you feel about this: 
* should I open a ticket / MR here at Directus to include some fixes, or 
* should I raise this in the OpenAPITools repo?  
* both..?

I am happy to provide a repro github for this issue so you can see what I mean.

**Comment by rijkvanzanten** [comment]
Do those errors indicate a problem in the exported types from Directus, or is it a compatibility issue between the OpenAPI spec we generate and that particular package? 🤔 I've never used that particular generator, so I'm not really sure what i'm looking at

**Comment by CoenWarmer** [comment]
I've created a repro that hopefully describes what I mean: https://github.com/CoenWarmer/directus-spec-openapi-repro/

I've written down what things I adjusted in the spec to make the generator work a bit better.

---

_Some_ of the issues stem from the spec that Directus provides (such as having `/components/definitions` which apparently is against the Open API schema) 🤷🏻‍♂️.

Others may be due to this particular generator; which is where my hesitance in opening a ticket at Directus comes from.

**Reply by CoenWarmer** [reply]
@rijkvanzanten out of interest: which generator do you use instead?

**Reply by rijkvanzanten** [reply]
Don't really had to use them before myself! 

We should do a bigger sprint on the types stuff to include https://github.com/directus/directus/pull/8531 (which would be very useful for your use case as a matter of fact), https://github.com/directus/directus/issues/14208, https://github.com/directus/directus/issues/11358, and https://github.com/directus/directus/issues/6084. All pretty doable improvements, but all have been a bit backlogged as it's hard to prioritize them over other bigger bugs/features 🤔 (@licitdev, maybe we can plan out a little type generation sprint? We can discuss thursday 7/21)

**Reply by CoenWarmer** [reply]
The exporting types feature looks amazing; would have to think about how we would deal with the endpoints, as one of the nice features of OpenAPIClient Generator is that you get full type annotated functions for the whole API which you can then easily include into React Query or whatever app you're building. 

On topic though: it seems as though the `XMetadata` and `Query` parts in the Directus spec currently cause this generator to generate faulty client code; when moving them to `/components/schemas` it does work.

Do you agree that I can open issues for them here? Or do you want me to take it to the OpenAPIGenerator project?

**Reply by rijkvanzanten** [reply]
Sure thing! If that's a spec problem in the directus output, then by all means feel free to open a new issue and we can triage/resolve it form there 🙂
